### DISCUSSION TITLE
Integration of sandboxed-api

### DISCUSSION THREAD

**Discussion by valoq** [curl/curl#15539] [Ideas]
The [sandboxed-api](https://github.com/google/sandboxed-api/) project provides libraries the means to implement a separate api that allows to execute library functions in a separate process which can be effectively restricted using seccomp and other means.
This is in essence the same broker-sandbox architecture as it is also implemented in chromium and firefox, resulting in severe security improvements by limiting the consequences of vulnerability in parts of their applications.

While looking into this project, I noticed there is already an implementation for curl available: https://github.com/google/sandboxed-api/tree/main/oss-internship-2020/curl
Since there is no sign of any pull requests to integrate this into the official curl project, I was wondering if this has ever been considered or if there are reasons why this has not been attempted.

**Comment by dfandrich** [comment]
I don't recall anyone proposing that before. I don't know much about sandboxed
API beyond what you've said, but that GitHub project's notes seem to indicate
that the libcurl API needs to change in some fundamental ways for it to work.
That doesn't sound like something that libcurl is likely to do.

What is the matter with the current approach, i.e. just using that project?
It's very much like a language binding, wrapping libcurl in a way so that the
"language" (users of simple-api) is able to use it. libcurl doesn't support any
of the other language bindings, and they all happily co-exist as well.

**Reply by valoq** [reply]
The way I understand it, the sandboxed curl api would be an alternative to choose from by developers using libcurl. Therefore it should, in theory, create no issues with anyone who does not choose to use this feature.
I did not yet experiment with the linked implementation, but it would make sense to me to find this as part of the official project as that would also add some confidence about the support of such an api.

**Comment by bagder** [comment]
@valoq, do you want to use this kind of API to libcurl in a libcurl using application of yours? If not, then I think this is not so interesting as no one has asked for this. If you *are* looking forward to using this in your application, then the question is still as @dfandrich asked: why does it have to done be "native" libcurl and not just in a binding on top of the library? It seems much more suitable as a binding on top since then the ones who want it can opt-in and those who don't want it, don't have to get bothered by it.

**Reply by valoq** [reply]
Perhaps you are correct that it is fine to use this api outside of this repo, though I would assume that integrating it directly into libcurl would help promote this api as an alternative for security focused applications.
I brought this up because am considering to integrate the sandboxed-api approach in another library and while researching it I noticed the curl implementation (among others) of that api and wondered if there is a reason why nobody is touching this. Seems I did overestimate the potential interest from users though.

**Reply by bagder** [reply]
>  I would assume that integrating it directly into libcurl would help promote this api as an alternative for security focused applications.

Do you want to use this kind of API to libcurl in a libcurl using application of yours? I don't think we need to second-guess what other users want. Exactly no users have asked for this.

> I noticed the curl implementation (among others) of that api 

It's not even a half-baked attempt. It's like a silly first step that is far from close to becoming usable.

**Comment by valoq** [comment]
On a related topic: Would you be interested in a MR that integrates this api into the curl cli tool? I was considering to play around with this existing libcurl api anyway and it should provide a useful option if it works without issues. This could be hidden behind a --enable-sandbox option that switches to the alternative api.

**Reply by bagder** [reply]
I don't see the point. I don't see the demand. I think you underestimate the work needed to make a useful implementation.

**Reply by valoq** [reply]
You are probably right about the work required and if there is no demand we can leave it at that.

If you don't mind I would appreciate if you could elaborate a bit on your own view towards features like this. You say you do not see the point in adding this and considering your position as a maintainer of a critical software component for probably most of the internet I would like to know more about that reasoning. 
If implementing and maintaining a feature like sandboxed-api does not appear as a good time investment in contrast to other task in your project, that would be an understandable point. I am curious though how you value the potential vulnerability impact reduction by this feature for curl. Demand by users may not be visible and even without any feature requests the benefits of containing much of the more complex code in a restricted runtime. Comparable security measures in OpenSSH and chromium have had a significant impact on the severity of vulnerabilities and curl appears to me as in a similar exposed position. Wouldn't it therefore make sense to have similar features in curl or am I overlooking something here?

**Reply by bagder** [reply]
Doing such isolation for (long-running) libcurl application use could potentially limit the scope of a certain kind of security issues, yes. But in order to do so you have to basically introduce a LARGE new set of API properties that no existing application uses, so it is expensive in every area: to implement and test, to adapt to for applications, to document and it will seriously increase foot-print of the product and it will perform worse.

But if you have unlimited time and resources, feel free to make an implementation and run tests and we can see for ourselves.

**Reply by valoq** [reply]
Thanks for the reply.
While unlimited time and resources are hard to come by, I think there may still be benefits of a partial api implementation for applications that use only some functionality and face a significant risk that justifies the efforts.

**Comment by dfandrich** [comment]
There are countless security add-ons that curl could integrate, but we can't
support them all (without undue cost on many vectors). See for example the
very similar discussion about a completely different solution at
https://github.com/curl/trurl/pull/278

**Reply by valoq** [reply]
Interesting reference, thanks.
I would place the sandboxed-api approach more as a framework then an alternative to the mentioned landlock implementation. There is still an open feature request for sandboxed-api to add landlock but in theory this api approach should make implementation maintenance easier then trying to build sandbox features from scratch. There is also some movement to create documentation for the entire approach https://github.com/ossf/wg-best-practices-os-developers/issues/631
